### DISCUSSION TITLE
OSX capture

### DISCUSSION THREAD

**Discussion by AmbiMod** [awawa-dev/HyperHDR#74] [General]
Hi Awawa,

My friend, I'm very sorry if I didn't give you any news about the banana... but a lot of things came in my life.

I got a baby girl ( the first one ) and she took me all my free time. I lost my job too few weeks ago so now I have time 😔 ( I'm too expansive so they decided to let me go away, nevermind ).

I didn't test anything on banana, I will see if I can do something maybe later. Now I have another problem, I bought a MacBook Pro M1 😁 ( Mac fan awawa ? )

Nice, but I want to put an Ambilight on it and it's a little bit complicated. Hyperion NG only works with the old generation of Mac and I need software capture.
I tried to compile it but I got more than 20 errors so I gave up.

Your HyperHDR version is working on the new Mac but, as you know, we cannot capture the desktop and this is my request.

Can you provide a HyperHDR version for Mac OS with this features ?

I use Prismatik on it for the moment but I don't have all the features provide by Hyperion ( black bar detector for example ).

Your works is really appreciate on the French Ambilight Discord, everybody install your Hyperion version.✌🏼

I hope you can do something for me. See you ☺️

**Comment by awawa-dev** [comment]
Hi
Thank you for the kind words and congratulations on the arrival of your new baby girl! 
Can confirm that I've great feedback from many users from French Ambilight Discord also on my private channel: best regards to you all :-) 

I wasn't a dedicated fan of macOS (because I didn't use it) but  it happened coincidentally that I have opportunity to access some Apple platforms at the time. So I decided to take that chance to jump with HyperHDR on that system. The more I work the more I like it: at least at the c++ brew level ;-)

As for a screen capture: I don't say no but it's rather a distant feature. You know my attitude towards software screen capturing ;) but that may a little change because I have some ideas that could use it. Anyway near term solution is probably to use some kind of virtual camera that would provide screen capturing stream. Tell me: did you try OBS?

**Reply by AmbiMod** [reply]
Thank you for your sweet words Awawa.☺️

I already test the OBS way and manycam too but it's not working. Maybe I do something wrong.

With OBS impossible to get a capture of the desktop and with manycam, I got a capture but not in HyperHDR.

I will try another time if I can have something but I'm not optimist.

Good to hear that your brain is always working to give us new tools and functionality 😁

I keep you informed 😉

**Reply by awawa-dev** [reply]
If OBS won't work please upload HyperHDR's logs on pastebin or similar service (most interesting is just after the start when HyperHDR has access right to the video devices and then it lists all of available capture devices and video modes).
I guess some details prevent the OBS virtual camera from working because I limited macOS port to modes that are supported by the hardware USB grabbers in my possession. But we have aces in the sleeve and maybe there is some easy workaround. At least I hope there is ;)

**Comment by awawa-dev** [comment]
Hi @AmbiMod 
Did you try THAT magic option for screen capturing using Google Chrome browser on macOS? :)
![macos](https://user-images.githubusercontent.com/69086569/122464190-dafddd80-cfb6-11eb-9736-5504aa7f7dbc.jpg)

**Reply by AmbiMod** [reply]
Sorry Awawa, I didn't reinstall OBS to give you the log when I try to use it with HyperHDR. I will do it asap.

For the camera option, I'm very stupid because i talk about it in my tutorial. 😖

![Capture d’écran 2021-06-18 à 12 33 20](https://user-images.githubusercontent.com/26428901/122548257-67e57d00-d031-11eb-9906-5f9d10bc72c7.jpg)

As you can see, it's working as expected.

![Capture d’écran 2021-06-18 à 12 19 14](https://user-images.githubusercontent.com/26428901/122547612-9e6ec800-d030-11eb-9cfc-3fd57ac58090.jpg)

It's not the same but I will use it like that, it's better than use Prismatik. Thanks for the trick recall. 😁

I hope you can add the desktop capture features in your next release.🦾

Take care. 😊

**Comment by awawa-dev** [comment]
Yes, it happens to me to forget about that grabber too because of the strict browser security policy it is not always available ;)
You must access the service as localhost or remotely using https instance on 8092 default port. On http 8090 it won't work. It works best with Google Chrome.

Now I can confirm that system grabbers are coming: one per each system, written from the scratch.
But as the version that is currently under development already brings huge amounts of changes (and I mean it's REALLY huge) it will be split into: alpha, beta and then stable release. I hope you will join the testing after it becomes available :)

**Reply by AmbiMod** [reply]
Nice to hear that a good things will come and of course, you can count on me for testing your future release.

Thank you for the additional information, I will see what I can do. I need to change the Arduino Nano I used with Prismatik by an ESP ( I will use it as Adalight device ).

BTW, I purchased a new USB3 capture device, an avaccess 4KVC00. Unfortunately it's not working with HyperHDR.
https://www.avaccess.com/products/4kvc00

I tested it on Windows and I will test it on Mac and Raspberry ( but I'm not optimist ). I saw a post on Hyperion Project about this device but I wanted to try by myself what's happen. 
https://hyperion-project.org/forum/index.php?thread/11605-usb-3-0-hdmi-video-capture-4kvc00-not-working/

Do you think you can do something to make it working if I give you a log ?

Thanks

**Comment by awawa-dev** [comment]
Can't read that forum's log for sure: it seems it's a mix of Macrosilicone (idVendor=534d; idProduct=2109, rather trouble free if you avoid 1080p capture mode) and some other device 1e4e:7016. Don't know if there were two capture devices connected at the same time or (1e4e:7016)  is a parent of some subclasses devices including that Macrosilicone. Must check it more detailed on Linux how it looks. If there are multiple devices on one bus line then you need a really good power supply from the USB port exceeding standard 500ma from USB2.0. And hoping that Windows/Linux kernel drivers can find their way to that nested Macrosilicone grabber. Logs from Windows and Rpi would be helpful also.

**Reply by AmbiMod** [reply]
Ok, i will give you the log I got on the three architecture, Windows, Mac and Raspberry.

I can give you the log I got with OBS on the Mac. You can find it here https://pastebin.com/raw/RsWuX5qz

![Capture d’écran 2021-06-18 à 21 05 44](https://user-images.githubusercontent.com/26428901/122606300-00532000-d079-11eb-8ee5-14591d636504.jpg)

I tried to capture the desktop, an image... but nothing. Here a screenshot of a movie captured by OBS.

Maybe this message are helpful, I don't know

OBS Virtual Camera (obs-virtual-cam-device) 1280 x 720 @ 30 fps 2VUY (unsupported)

**Reply by awawa-dev** [reply]
2VUY is not supported in HyperHDR...yet ;-) If think I can add it without any problem. Don't remember for sure but maybe it was even present in some early version of HyperHDR and then was decided that it's legacy, not used anymore video format (and it's true for USB hardware devices).

**Reply by awawa-dev** [reply]
BTW you must test that  4KVC00 device using USB3 so for example RPi3 is out. I think the problem could be power supply limit and not slow speed of USB2.0.

**Reply by AmbiMod** [reply]
I will try with my pi4. The device is auto powered by the USB connected on the USB of the PC/RPI...Maybe a wrong thing. I will see if I can add something this week end on raspberry.

Apparently the 4KVC00 use YUY2, you can add it too ? Or maybe it's not the problem with this device 😒

**Reply by awawa-dev** [reply]
YUY2 is supported already. It's a standard color format.
Try to lower the resolution and see if it helps. It maybe related to some problems with setting video format so HyperHDR force option could come in handy.

**Reply by AmbiMod** [reply]
Ok i will check. Maybe tomorrow I will purchased a chromecast and see what I got.

I will keep you informed after all my test. 

thanks again. 😉

**Reply by awawa-dev** [reply]
One more thing: if you have pc or notebook with USB 3.0 type C test it using type C to type C cable.
I have a long struggle with my Ezcap 320 and it refuses to work in USB3.0 mode if it is not connected using type C to type C.
Type C to type A is not enough. Maybe that's the case?
PS:
At least they fixed other issues with Ezcap 320 :)

**Reply by AmbiMod** [reply]
> BTW you must test that 4KVC00 device using USB3 so for example RPi3 is out. I think the problem could be power supply limit and not slow speed of USB2.0.

It's working on Mac 😁

![Capture d’écran 2021-06-20 à 10 27 31](https://user-images.githubusercontent.com/26428901/122667364-c71ecb00-d1b2-11eb-9f13-b2e627291ead.jpg)

I will test on raspberry now and I hope I got the same 🙏🏼

**Reply by AmbiMod** [reply]
and on raspberry too

![Capture d’écran 2021-06-20 à 12 27 52](https://user-images.githubusercontent.com/26428901/122670578-179e2480-d1c3-11eb-893b-e3240afb953d.jpg)

But you right about the power, I need 3A for the PI and the Loop. I think I will buy an Y cable to give more power to the loop.

**Comment by AmbiMod** [comment]
Just another message.

I have a "problem" with the ressource of my pi when I capture with this device

As you can see, all at 100%.

![Capture d’écran 2021-06-20 à 13 55 59](https://user-images.githubusercontent.com/26428901/122673353-571f3d80-d1d0-11eb-8ec9-83bc90c9a1db.jpg)

I need to check this box to reduce the amount of % and have something smooth.
![Capture d’écran 2021-06-20 à 13 56 18](https://user-images.githubusercontent.com/26428901/122673362-63a39600-d1d0-11eb-90c3-5ca9b15d632d.jpg)

![Capture d’écran 2021-06-20 à 13 56 11](https://user-images.githubusercontent.com/26428901/122673360-60a8a580-d1d0-11eb-852c-19344081afaa.jpg)

Oh, and good things, no tv pattern with this box 😍 so exit rainbow or green picture.

**Reply by awawa-dev** [reply]
The device doesn't offer low resolution? I would not use custom video resolution: choose the video resolution from the list if there are available options. Custom resolution can't offer nothing more than that and it's already removed from the developing version. 
Also make sure that you are running yuyv encoding in the logs. Force mode works only if the device has such capability.

**Reply by AmbiMod** [reply]
Nope, Only one choice in the list, 1980x1024 and nothing else.

I also use the yuyv encoding. 😊

Bad News, this device  is not CEC as expected but with help of a friend on the discord, we can bypass the problem by doing this 
![image0](https://user-images.githubusercontent.com/26428901/122677268-f13bb180-d1e1-11eb-9c3b-a323cee5aed9.jpg)

**Reply by awawa-dev** [reply]
Precision work :)

Anyway CPU usage indicates that the grabbers is using MJPEG stream despite your settings (yuyv is probably unavailable).

**Reply by AmbiMod** [reply]
Hi awawa, 🙂

I tested HyperHDR on my Mac with the desktop sharing features provide in Google Chrome but i have a lot of latency, near 0.5s

I tried many Adalight sketch on my ESP but the result is always the same :
- WLED wifi
- WLED using your Adalight protocol
- Your Adalight protocol directly
- An Adalight ino I found when I used prismatik

I also tried this Adalight sketch https://raw.githubusercontent.com/sticilface/Adalight-ESP8266/master/Adalight-ESP8266.ino that I modified for RGBW but I have trouble with colors when I captured the desktop ( I got the good colors, RBG test sequence, at boot when I plug the ESP but not in capture, white Is good but Christmas ledstrip the rest of the time, nevermind ).

I changed all the settings to optimize HyperHDR but nothing works. The problem is the capture method I guess.

I can't use HyperionRBG V2 too ( HyperionRGB version modified by a friend of mine from the discord https://github.com/provalinf/HyperionRGB ) because RGBW is not implemented, only RGB Ledstrip, damn.

But I don't think it will change the problem.

Do you hear from another one using this function on Mac that everything is working correctly ?

**Reply by awawa-dev** [reply]
Hi
I've got only feedback when HyperHDR is used with USB grabber and I tested only that combination on macOS. But the streamer performance should be rather system independent.
You can tune up a bit streamer but cant guarantee it will improve the situation much. Or you can test Firefox for comparison.
As you can check in streamer.js file, screenshotIntervalTimeMs is set to 100 , and video framerate is requested as 15FPS.
With these settings in worst case scenario: base latency is 100ms +  65ms = 165ms. But it's rather far from 0.5 latency you experience.

**Reply by AmbiMod** [reply]
Thanks awawa, i will see what i can do for the mac.

I have a question :

What do you think about this ?
![4](https://user-images.githubusercontent.com/26428901/123481504-1e35fd00-d604-11eb-9ab0-ec93af861daa.jpg)
![5](https://user-images.githubusercontent.com/26428901/123481512-1fffc080-d604-11eb-8d34-5342d3c6a310.jpg)

![1](https://user-images.githubusercontent.com/26428901/123481265-cac3af00-d603-11eb-9c63-e090a8a28fc9.jpg)
![2](https://user-images.githubusercontent.com/26428901/123481284-cf886300-d603-11eb-8873-92770c674722.jpg)
![3](https://user-images.githubusercontent.com/26428901/123481302-d44d1700-d603-11eb-9786-34730f3e3063.jpg)

I use my ambilight playing HDR and non HDR content with HDR to SDR box uncheked using my 4KVC00 loop.

**Reply by awawa-dev** [reply]
Did you check the real video stream format in the logs? I'm curious because high CPU says it's rather MJPEG video stream. On Rpi4 1080p yuyv should be around 200-250% CPU and mjpeg above 350% . And for SDR/HDR if the colours are in fact not washed out then maybe it does some kind of tone mapping like Ezcap 320? Although comparing picture 3 & 4 suggests it's rather without hardware tone mapping.

BTW I built ws2813 led frame that I mentioned you before. Had to build new professional level shifter based on 74AHCT125 also because I swap different frames/leds for my tests. ws2813 has definitely better blue component that is the weakness of sk6812 and more vivid bright colors, but in the normal or dark scene the colors are over-saturated and I dont like it. If I only had one white wall in my home then I would record a video to show the difference. So I returned back to my sk6812 cold white.
